text/0000-return-type-notation.md

```diff
@@ -0,0 +1,1105 @@
+This formulation can be done with `trait_variant` like so
+
+```rust
+#[trait_variant::make(Service: Send)]
+trait LocalService<R> {
+    type Response;
+    
+    async fn call(&self, request: R) -> Self::Response;
+}
+```
+
+However, it carries the downside that users must implement `LocalService` and hence must be aware of the desugaring.
+It would be nicer if users could choose to implement `Service` and then (in so doing) effectively assert that all their async functions are *always* `Send`.
+This is not possible today due to the fact that `trait-variant` is emulating trait alias functioanlity with a blanket impl and supertraits; this is because true trait alias functionality is not yet stable.
+[RFC 3437][] has proposed an extension to trait aliases that makes them implementable.
+The combination of accepting [RFC 3437][] and stabilizing trait aliases would make these aliases nicer for users as a result.
+
+[RFC 3437]: https://github.com/rust-lang/rfcs/pull/3437
+
+## Permit RTN for more functions
+
+RTN is currently limited to `async fn` and `-> impl Trait` methods in traits.
+But the same syntax could be used for any methods as well as for free functions (e.g., `foo(..)` might refer to the return type of `fn foo()`).
+One area that would be challenging to support is RTN for the return types of closures,
+as that would introduce an element of dependent types that would complicate the type checker
+(e.g., if `let y: x(..)` meant that `y` is the type returned from invoking the closure `x`, another local variable).
+
+## Specifying the values for argument types
+
+The `T::method(..): Send` notation we've been using so far means
+"the return type of `method(..)` is `Send`, no matter what arguments you provide".
+We could extend this notation to permit specifying the argument types explicitly.
+For example, consider the `capture` method below, which takes a parameter of type `input`:
+
+```rust
+trait Capture {
+    async fn capture<T>(&mut self, input: T) -> Option<T>;
+}
+```
+
+and now consider a function that invokes `capture` with an `i32` value:
+
+```rust
+async fn capture_i32<C: Capture>(mut c: C) {
+    c.capture(22_i32);
+}
+```
+
+Now imagine we wanted to invoke `capture` on another thread,
+and hence we need a where-clause indicating that the future
+returned by `capture` will be `Send`:
+
+```rust
+async fn capture_i32<C: Capture + Send + 'static>(mut c: C)
+where
+    /* where-clause for C::check() needed here! */
+{
+    workstealing_runtime::spawn(async move {
+        c.capture(22_i32);
+    })
+}
+```
+
+There are multiple ways we could write this where-clause, varying in their specificity...
+
+* `where C::capture(..): Send` -- this indicates that `C::capture()` will return a `Send` value for any possible set of parameters
+* `where C::capture(&mut C, i32): Send` -- this indicates that `C::capture()` will return a `Send` value when invoked specifically on a `&mut C` (for the `self` parameter) and an `i32`
```

Comment:
wouldn't this bring back the ambiguity with `Fn()`? for example when we have `fn capture() -> impl Future`

Reply:
Indeed, it would. `fn f() where C::capture(&mut C, i32): Send {}` is already syntactically legal *today*. The parentheses denote *parenthesized type arguments* which may be attached to arbitary type paths (and it also comes with turbofish for expression and pattern paths: `C::capture::(&mut C, i32)`).

Reply:
It is syntactically legal, yes, but not meaningful semantically.

Reply:
Semantically as well if `C` is a module but yeah, I don't think it can actually meaningfully clash if `C` is a type parameter.

```rs
//@ edition: 2015
//@ check-pass
mod C { pub use std::ops::Fn as capture; }
fn f() where C::capture(): 'static {}
```